### PR TITLE
Make gpuCI and pre-commit style configurations consistent

### LINKED ISSUE #8193: [FEA] Make style CI and pre-commit hooks consistent
**Is your feature request related to a problem? Please describe.**
Running `pre-commit run --all-files` against the codebase results in various test failures, while the style CI reports these same tests as passing.

Some of these are related to the pre-commit hooks checking files that are typically ignored (such as [`ci/utils/nbtestlog2junitxml.py`](https://github.com/rapidsai/cudf/blob/branch-0.20/ci/utils/nbtestlog2junitxml.py)), while others seem to be related to the fact that the config used when running the hooks is different from that used in the style CI (namely `isort`, `black`, and `clang-format`).

**Describe the solution you'd like**
Either:

- Changing `ci/checks/style.sh` to enforce the additional rules set by the pre-commit hooks
- Changing the various configuration files used by pre-commit so that the checks pass on the codebase as is

I imagine the latter option would be preferable as it would avoid having to run the pre-commit hooks against the codebase in a large PR.

**Describe alternatives you've considered**
Not doing anything; this is purely a QOL thing and can be entirely ignored since the pre-commit hooks don't seem to outright conflict with the style CI.

### PR DESCRIPTION
**Pull request by charlesbluca**
Closes #8193

Makes changes to the pre-commit config, gpuCI style script, and `cpp/scripts/run-clang-format.py` so that the styling of the pre-commit hooks should roughly match the style checks done by gpuCI. In particular:

- pre-commit's black and flake8 hooks now only target `python/` to match with gpuCI
- gpuCI's `cpp/scripts/run-clang-format.py` now runs `clang-format` on `cpp/libcudf_kafka/` and `java/src/main/native/` to match with pre-commit (also removed `cpp/include/cudf` and `cpp/include/nvtext` as they were redundant here)
- isort has been split up by project (cudf, dask-cudf, etc.) and now runs on each project using its respective configuration; this was done because currently the isort check on gpuCI was only checking Python files one level deep in each project

It would probably be good to run the updated hooks on the codebase either in this PR or a linked one so that this changes doesn't break tests.

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (charlesbluca) [inline .pre-commit-config.yaml:8]: For whatever reason, this hook resorts `__init__.py` files, despite the config specifying to skip them; trying to figure out a way to actually skip them.

File: ci/checks/style.sh
Comment (charlesbluca) [inline ci/checks/style.sh:17]: For some reason, the pre-commit hooks for `isort` automatically skip Cython files, while the CLI does not, so I had to specify to skip them here. Do we want package sorting for Cython files?

**Review by charlesbluca — COMMENTED** [review]
Some minor quirks around the behavior of `isort`:

**Comment by mythrocks** [comment]
> gpuCI's `cpp/scripts/run-clang-format.py` now runs `clang-format` on `cpp/libcudf_kafka/` and `java/src/main/native/` to match with pre-commit.

The fact that this runs on `java/src/main/native/` makes me a tad nervous. Not to imply that I disagree; I would prefer the formatting be applied too. It's just that this implies a fair bit of formatting changes to the existing JNI/native code. 
cc @revans2, @jlowe.

**Comment by mythrocks** [comment]
I think this PR would need to include the results of running `run-clang-format` on `cpp/libcudf_kafka` and `java/src/main/native`. For the moment, we should include this as a separate commit, so as not to muddy the review.

Also, @charlesbluca, could you please confirm that as a result of this commit, the intention is for  clang formatting to run on `cpp/libcudf_kafka` and `java/src/main/native`, even for PRs where `[skip-ci]` is specified?

I ask because `java/src/main` tests (even the `native` ones) are not currently run as part of CI for historical reasons. These PRs (e.g. #8209) are submitted with `[skip-ci]` so as not to use up CI resources in rerunning unrelated `libcudf` tests. 

It would be good to have `[skip-ci]` *not* skip `clang-format` checks in `libcudf_kafka` and `java`. Otherwise, the errant code would get merged, and fail in other PRs that don't `[skip-ci]`.

**Comment by charlesbluca** [comment]
> For the moment, we should include this as a separate commit, so as not to muddy the review.

I agree - I was planning to run through the pre-commit hooks once all the config changes were ironed out.

> the intention is for clang formatting to run on cpp/libcudf_kafka and java/src/main/native, even for PRs where [skip-ci] is specified?

No, the intent is so that when gpuCI is ran, it will run `clang-format` on `cpp/libcudf_kafka` and `java/src/main/native`, in addition to all the other (in your case, unnecessary) tests. What you're describing does sound like a good idea though, but I'm not really sure how gpuCI handles conditionals for testing (i.e., only run this set of tests if these files are impacted by a PR).

A relatively simple workaround for this which would address the situation you described and also make this PR a little simpler would be to make a GitHub Actions workflow running a [pre-commit action](https://github.com/pre-commit/action) (or using [precommit.ci](https://pre-commit.ci/), which seems to be its successor). This would ensure that the style checks being ran when making commits and opening PRs are always the same, since they rely on identical configs; it would also create a workflow to check for formatting independent of if a PR is submitted with `[skip ci]`.

That being said, I'm not sure of the potential downsides to moving aspects of our testing outside of gpuCI.

File: ci/checks/style.sh
Comment (ajschmidt8) [inline ci/checks/style.sh:30]: please make sure to add all of the new `ISORT` variables to the line linked below

https://github.com/rapidsai/cudf/blob/fb7cdcd7bc538298d4df684c6c7747cf5fbf3c70/ci/checks/style.sh#L106

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: ci/checks/style.sh
Comment (charlesbluca) [reply ci/checks/style.sh:17]: Looks like we can get Cython/PYI resorting in pre-commit's isort hook with:

```yaml
types: [text]
types_or: [python, cython, pyi]
```

So if we would like package resorting on Cython it's now totally doable.

File: cpp/scripts/run-clang-format.py
Comment (hyperbolic2346) [inline cpp/scripts/run-clang-format.py:32]: did you intend to remove `cpp/include/cudf` here? I didn't see that called out anywhere.

Comment (charlesbluca) [reply cpp/scripts/run-clang-format.py:32]: If I'm not mistaken, I believe this directory should still be getting checked because parent directory `cpp/include` is listed above - I made this change (along with removing `cpp/include/nvtext`) out of clarity since listing the two subdirectories seemed redundant.

Comment (charlesbluca) [reply cpp/scripts/run-clang-format.py:32]: That being said, I did forget to call that out specifically in the PR body - if you're okay with this change, I'll go ahead and add it there.

Comment (hyperbolic2346) [reply cpp/scripts/run-clang-format.py:32]: My only concern is that we're still doing what we expect. I don't have an issue if this is recursive as it is redundant.

**Review by hyperbolic2346 — APPROVED** [review]

**Comment by charlesbluca** [comment]
To update the separate style CI issue, rapidsai/ops#1580 has been opened - @mythrocks, would you prefer that get resolved before merging this?

**Comment by harrism** [comment]
@charlesbluca what needs to happen to get this in? We are entering code freeze tomorrow.
@mythrocks can you answer the question above?

**Comment by charlesbluca** [comment]
I think this would be better retargetted to `21.08`; the C++ formatting is generally sorted out, but I would like more of a conversation around if we want to do package sorting for Cython code. Other than that, we would need to run the updated hooks against the codebase and make a lot of formatting changes, which we probably don't want to do this close to release.

**Comment by mythrocks** [comment]
Sorry I missed this question. Please go ahead with this change. 
I wanted to avoid cases where `skip-ci` PRs were checked in, and broke unrelated PRs for others. We can address that later, if that's more acceptable.

**Review by mythrocks — APPROVED** [review]

**Comment by charlesbluca** [comment]
Thanks @mythrocks! Now the only blocker on this is rapidsai/integration#286, which would allow us to resort packages in .pxd files.

**Review by charlesbluca — COMMENTED** [review]
With these changes, `isort` should now check and resort cython + pyi files, and once rapidsai/integration#286 is merged we can safely merge this; marking this as draft until then

**Comment by charlesbluca** [comment]
The style failures here seem to be due to an `isort` version mismatch - I can confirm that the gpuCI script passes locally with `isort=5.6.4`.

A workaround here could be to skip over the impacted files (cuDF pyx/pxd files and `python/dask_cudf/dask_cudf/io/tests/test_json.py`) in this PR, and retarget those files in a follow up PR once rapidsai/integration#286 is merged

**Comment by charlesbluca** [comment]
To be specific on the failure that happened in `python/dask_cudf/dask_cudf/io/tests/test_json.py`, it looks like `isort=5.0.7` doesn't allow imports like:

```python
import dask.dataframe as dd
```

instead preferring:

```python
from dask import dataframe as dd
```

Whereas `isort=5.6.4` generally allows either of these import styles. Do we have a preference for which style we want to enforce? There might be a setting which forces imports to be one or the other.

**Comment by galipremsagar** [comment]
> Whereas `isort=5.6.4` generally allows either of these import styles. Do we have a preference for which style we want to enforce? There might be a setting which forces imports to be one or the other.

I'd suggest to keep it simple as:
```python
import dask.dataframe as dd
```

**Comment by vyasr** [comment]
fwiw `import dask.distributed as dd` is also faster than `from dask import distributed as dd`. Generally a negligible microoptimization, but for function-local imports that can matter.

**Comment by charlesbluca** [comment]
I couldn't find an option to force that specific import style so I just skipped `isort` for that specific import; once rapidsai/integration#286 is merged, I can remove that skip

**Comment by galipremsagar** [comment]
Can we update this PR now that integration changes are in?

File: cpp/libcudf_kafka/tests/kafka_consumer_tests.cpp
Comment (galipremsagar) [inline cpp/libcudf_kafka/tests/kafka_consumer_tests.cpp:24]: Can you update copyright year in this file?

File: cpp/scripts/run-clang-format.py
Comment (galipremsagar) [inline cpp/scripts/run-clang-format.py:29]: copyright

File: java/src/main/native/src/HostMemoryBufferNativeUtilsJni.cpp
Comment (galipremsagar) [inline java/src/main/native/src/HostMemoryBufferNativeUtilsJni.cpp:19]: copyright.

File: java/src/main/native/src/NvcompJni.cpp
Comment (galipremsagar) [inline java/src/main/native/src/NvcompJni.cpp:29]: copyright.

File: java/src/main/native/src/NvtxRangeJni.cpp
Comment (galipremsagar) [inline java/src/main/native/src/NvtxRangeJni.cpp:22]: copyright.

File: python/cudf/cudf/_lib/cpp/strings/extract.pxd
Comment (galipremsagar) [inline python/cudf/cudf/_lib/cpp/strings/extract.pxd:1]:
```suggestion
# Copyright (c) 2020-2021, NVIDIA CORPORATION.
```

**Review by galipremsagar — CHANGES_REQUESTED** [review]
Can you update the copyright year for all the files?

File: python/dask_cudf/dask_cudf/io/tests/test_json.py
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/io/tests/test_json.py:11]: Why the override here?

**Review by rjzamora — APPROVED** [review]
Dask-cudf changes seem fine - Just left a minor question

File: python/dask_cudf/dask_cudf/io/tests/test_json.py
Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/io/tests/test_json.py:11]: Now that gpuCI's `isort==5.0.7` check is targetting this file, it want to change this line to

```python
from dask import dataframe as dd
```

I couldn't find an option for `isort` to allow this type of import, but can confirm that `isort==5.6.4` (which we are planning to bump gpuCI to after this PR is merged) allows either style of import, so this override can be removed once rapidsai/integration#286 is merged.

**Review by galipremsagar — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Review by jlowe — APPROVED** [review]
Java approval

**Comment by charlesbluca** [comment]
@gpucibot merge

**Comment by charlesbluca** [comment]
After playing around with Cython resorting for a while, I noticed a quirk of `isort` means we probably aren't seeing nearly as much resorting as we would if our pyx/pxd files used exclusively `cimport` statements. I've opened up an issue describing this in PyCQA/isort#1787, so that hopefully we can make some modifications to our `setup.cfg` files to account for this.

**Pull request merged**
